Add ParamsValidator

AndreyAkinshin opened the pull request.

timcassell commented on src/BenchmarkDotNet/Validators/ParamsValidator.cs:
```diff
@@ -0,0 +1,31 @@
+        {
+            var constantsWithParams = type
+                .GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.FlattenHierarchy)
+                .Where(fieldInfo => fieldInfo.IsLiteral && !fieldInfo.IsInitOnly && fieldInfo.HasAttribute<ParamsAttribute>());
```
```dif
- fieldInfo.IsLiteral && !fieldInfo.IsInitOnly
+ fieldInfo.IsLiteral || fieldInfo.IsInitOnly
```

We should check for both const and readonly.

timcassell replied:
Also should check for `ParamsAllValuesAttribute` and `ParamsSourceAttribute` (unless you want to duplicate this check in the `ParamsAllValuesValidator` and add a `ParamsAllValuesValidator`).

timcassell commented:
Do you want to include checks for #1877 in this validator?

AndreyAkinshin commented:
@timcassell thanks for the comments! I have supported readonly fields and added checks for multiple Params* attributes to fix #1877

timcassell commented:
Awesome! One more thing I think could help is making sure a property setter exists, and is public and not marked `init` or `required`.

adamsitnik commented on src/BenchmarkDotNet/Configs/DefaultConfig.cs:
```diff
@@ -68,6 +68,7 @@ public IEnumerable<IValidator> GetValidators()
             yield return GenericBenchmarksValidator.DontFailOnError;
             yield return DeferredExecutionValidator.FailOnError;
             yield return ParamsAllValuesValidator.FailOnError;
+            yield return ParamsValidator.FailOnError;
```
we should also add it to the list of mandatory validators:

https://github.com/dotnet/BenchmarkDotNet/blob/ccbaf08b03b38d1d418756a9e528c60c9f4951d1/src/BenchmarkDotNet/Configs/ImmutableConfigBuilder.cs#L17-L31

so it's applied even when users provide an empty config

adamsitnik reviewed (approved):
Overall it LGTM (great test coverage). Please add this validator to the list of mandatory validators before you merge the PR. Thanks!

AndreyAkinshin commented:
@timcassell @adamsitnik thanks for the review. I have addressed your comments:

* Added validation and tests for init-only properties (.NET5+ only)
* Added ParamsValidator to the list of mandatory validators
* Removed `[PublicAPI]` for `ReflectionExtensions` methods

timcassell commented on tests/BenchmarkDotNet.Tests/Validators/ParamsValidatorTests.cs:
```diff
@@ -0,0 +1,206 @@
+            [ParamsSource(nameof(Source))]
+            public bool Input { get; init; }
+        }
+
```
Add tests for `private set` and no setter?

timcassell commented on tests/BenchmarkDotNet.Tests/Validators/ParamsValidatorTests.cs:
```diff
@@ -0,0 +1,206 @@
+            [ParamsSource(nameof(Source))]
+            public bool Input { get; set; }
+        }
+
```
Add tests for `private set` and no setter?

AndreyAkinshin replied:
It's tricky. Currently, if no public setter is available, BenchmarkDotNet throws [this exception](https://github.com/dotnet/BenchmarkDotNet/blob/v0.13.5/src/BenchmarkDotNet/Extensions/ReflectionExtensions.cs#L181):
```cs
Unhandled exception. System.InvalidOperationException: Member "Input" must be public if it has the [ParamsAttribute] attribute applied to it
   at BenchmarkDotNet.Extensions.ReflectionExtensions.GetTypeMembersWithGivenAttribute[TAttribute](Type type, BindingFlags reflectionFlags)
   at BenchmarkDotNet.Running.BenchmarkConverter.<>c__DisplayClass11_0.<GetParameterDefinitions>g__GetDefinitions|0[TAttribute](Func`3 getValidValues)
   at BenchmarkDotNet.Running.BenchmarkConverter.GetParameterDefinitions(Type type)
   at BenchmarkDotNet.Running.BenchmarkConverter.MethodsToBenchmarksWithFullConfig(Type type, MethodInfo[] benchmarkMethods, IConfig config)
   at BenchmarkDotNet.Running.BenchmarkConverter.TypeToBenchmarks(Type type, IConfig config)
   at BenchmarkDotNet.Running.BenchmarkRunner.RunWithDirtyAssemblyResolveHelper(Type type, IConfig config, String[] args)
   at BenchmarkDotNet.Running.BenchmarkRunner.<>c__DisplayClass0_0`1.<Run>b__0()
   at BenchmarkDotNet.Running.BenchmarkRunner.RunWithExceptionHandling(Func`1 run)
   at BenchmarkDotNet.Running.BenchmarkRunner.Run[T](IConfig config, String[] args)
```
This happens before the validation stage.

I think it would be better to abandon such exceptions in favor of `Validator` checks (this exception was introduced in 2017). However, it deserves a separate PR (I can work on it once this PR is merged).

timcassell replied:
@AndreyAkinshin That's the case for `public bool Input { get; private set; }`? The code looks like it just checks if the property itself is public, not its setter.

AndreyAkinshin commented:
@adamsitnik @timcassell the second part of Params checks is ready: #2283
